### PR TITLE
Add options for header names and ttl transformation

### PR DESCRIPTION
**Pull request by jsmpereira**
Introduce options `headerLimit`, `headerRemaining` and `headerReset`
to support defining header names.
Introduce option `ttlTransform` to support modifying the ttl format.

With the introduced options it is possible to support the "RateLimit
Header Fields for
HTTP" (https://tools.ietf.org/html/draft-polli-ratelimit-headers-01)
Internet-Draft.

### CODE REVIEW

**Comment by yonjah** [comment]
Hi @jsmpereira .
Thank you for your PR.

Do you think you'll be able to update the docs as well ?

File: hapi-plugin/lib/plugin.js [hunk 1]
```diff
@@ -17,9 +17,13 @@ const optionsSchema = joi.object().keys({
 	bucket: joi.string().alphanum(),
 	countSuccess: joi.boolean().default(true),
 	addHeaders: joi.boolean().default(true),
+	headerLimit: joi.string().default('X-RateLimit-Limit'),
+	headerRemaining: joi.string().default('X-RateLimit-Remaining'),
+	headerReset: joi.string().default('X-RateLimit-Reset'),
 	message: joi.string().default('you have exceeded your request limit'),
 	onError: joi.func().arity(3),
 	getKey: joi.func().arity(1).default(getRequestIP),
+	ttlTransform: joi.func().arity(1).default(identityTTL),
```
Comment (yonjah) [inline hapi-plugin/lib/plugin.js]: I wonder if it might be better to just use an arrow func here -
`ttlTransform: joi.func().arity(1).default(ttl => ttl)`

**Comment by jsmpereira** [comment]
@yonjah Changes pushed.

Please feel free to comment on/adapt the docs text. I'm not too sure about the wording. You might have something different in mind.

On another note, I didn't touch the express middleware, but I believe similar modifications would be trivial. Would you like me to take a crack at that or?

**Comment by yonjah** [comment]
Hi @jsmpereira  Thank you for making the changes.

Docs looks great.

I was going to suggest to add this to the middleware as well if you want. I think it's going to be very similar to the changes to the hapi plugin.

I can see the bug in https://github.com/yonjah/ralphi/pull/37/files#diff-361bc93adabdc14470e1b770f075d79bL57 everywhere else it returns a milliseconds accurate time and there it truncates it to seconds, I think the simplest fix for now will be to change it to milliseconds as well (even though it's a bit over the top) you can leave it as is and I'll do this change once I merge this in.

Are you sending headers following the spec in the draft ? 

I'm thinking maybe after we merge this PR to change the default way Ralphi works to be as described in the draft. 
It is going to be a breaking change but it might be worth it to offer it out of the box

**Comment by jsmpereira** [comment]
> I was going to suggest to add this to the middleware as well if you want. I think it's going to be very similar to the changes to the hapi plugin.

I can take a crack at it.

> Are you sending headers following the spec in the draft ?

I hope so, to the best of my understanding :sweat_smile:. This is my route configuration:
```js
plugins: {
  ralphi: {
    bucket: 'oauth',
    headerLimit: 'RateLimit-Limit',
    headerRemaining: 'RateLimit-Remaining',
    headerReset: 'RateLimit-Reset',
    onError: function (request, reply, error) {
      request.log('error:rate-limit', error);

      return reply.continue();
    },
    getKey: function (request) {

      return request.auth.credentials.clientId;
    },
    ttlTransform: function (ttl) {

      return Math.floor(((ttl * 1000) - Date.now()) / 1000);
    }
}
```

> I'm thinking maybe after we merge this PR to change the default way Ralphi works to be as described in the draft.
> It is going to be a breaking change but it might be worth it to offer it out of the box

I would say :+1:  to that.

Hopefully the draft will become a standard, but it will take quite some time for API's to drop their custom headers (x-ratelimit- variations) and adopt it. In any case, my basic stance on these things is that if there is at least a draft out there, getting some eyes and discussion, whatever is in there is as good (hopefully better) as anything I can come up with for my own in-house approach to X. I do like to point developers to RFC's instead of some custom page on our developer docs :)

**Comment by yonjah** [comment]
Cool.

I think we can safely change it to the draft spec once we add this changes.
As it's going to be the only change anyone who can't upgrade his consumers will be able to still use v 3.x and anyway it will be very minor conifguration change to get it to work the old way

**Comment by yonjah** [comment]
I've merge this PR in #41 
I checked the Error ttl again and I think it's ok. Unix time is sent everywhere (Ralphi Client change it from milliseconds to seconds) so it should be the same as TTL response without an error.
I've added the call to `ttlTransform` on the error TTL as well.

As mentioned earlier I'd like to change the default behaviour to support the draft, but there are a lot of major upgrades to do to the dependencies that I think it's better to do everything in the same major release so it might take a bit longer to get that version out

**Comment by jsmpereira** [comment]
@yonjah Thanks for taking in the changes :)

I was wondering, and I think that is the reason I didn't just apply `ttlTransform` on the error TTL, is that on error you get seconds and on normal execution you get milisecs, that means you can't apply the same transform, to which I refered to in a previous comment:

> I also noticed that when there's an error reaching the Ralphi server and onError is not defined, the ttl (https://github.com/yonjah/ralphi/pull/37/files#diff-361bc93adabdc14470e1b770f075d79bL57) is defined in a different format (seconds, as opposed to UNIX timestamp elsewhere) which would require that a custom ttlTransform check which format is receiving before applying the transformation, not to end with inconsistent values.

Or am I missing something?

> As mentioned earlier I'd like to change the default behaviour to support the draft, but there are a lot of major upgrades to do to the dependencies that I think it's better to do everything in the same major release so it might take a bit longer to get that version out

:+1: I'll be glad to provide feedback if you'd be interested.

**Comment by yonjah** [comment]
> I was wondering, and I think that is the reason I didn't just apply `ttlTransform` on the error TTL, is that on error you get seconds and on normal execution you get milisecs, that means you can't apply the same transform, to which I refered to in a previous comment:

The server returns milliseconds timestamps to all responses but it is later truncated by the client -
https://github.com/yonjah/ralphi/blob/master/client/lib/client.js#L49

So i think you should get unix timestamps everywhere.

I'm not sure why I did it that way. I guess I wanted to have the flexibility on the client to handle the TTL format but it's probably not really useful and doesn't worth the confusion

> +1 I'll be glad to provide feedback if you'd be interested.

Thanks, that be great. I hope I'll have time to work on it properly in the next couple of weeks

**Pull request closed without merging**